service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/azure/container/CopyAzureStorageContainerDefinitionStep.java

```diff
@@ -0,0 +1,155 @@
+
+    // get the inputs from the flight context
+    var destinationResourceName =
+        FlightUtils.getInputParameterOrWorkingValue(
+            flightContext,
+            WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_NAME,
+            WorkspaceFlightMapKeys.ResourceKeys.PREVIOUS_RESOURCE_NAME,
+            String.class);
+    var description =
+        FlightUtils.getInputParameterOrWorkingValue(
+            flightContext,
+            WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_DESCRIPTION,
+            WorkspaceFlightMapKeys.ResourceKeys.PREVIOUS_RESOURCE_DESCRIPTION,
+            String.class);
+    var destinationWorkspaceId =
+        inputParameters.get(
+            WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+    var destinationContainerName =
+        inputParameters.get(
+            WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_CONTAINER_NAME, String.class);
+    var destinationResourceId =
+        inputParameters.get(
+            WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_RESOURCE_ID, UUID.class);
+    var destStorageAccountId =
+        flightContext
+            .getWorkingMap()
+            .get(
+                WorkspaceFlightMapKeys.ControlledResourceKeys
+                    .DESTINATION_STORAGE_ACCOUNT_RESOURCE_ID,
+                UUID.class);
+
+    ControlledAzureStorageContainerResource destinationContainerResource =
+        buildDestinationControlledAzureContainer(
+            sourceContainer,
+            destStorageAccountId,
+            destinationWorkspaceId,
+            destinationResourceId,
+            destinationResourceName,
+            description,
+            destinationContainerName);
+    ApiAzureStorageContainerCreationParameters destinationCreationParameters =
+        new ApiAzureStorageContainerCreationParameters()
+            .storageContainerName(destinationContainerName)
+            .storageAccountId(destStorageAccountId);
+    ControlledResourceIamRole iamRole =
+        IamRoleUtils.getIamRoleForAccessScope(sourceContainer.getAccessScope());
+
+    // create the container
+    ControlledAzureStorageContainerResource clonedContainer =
+        controlledResourceService
+            .createControlledResourceSync(
+                destinationContainerResource, iamRole, userRequest, destinationCreationParameters)
+            .castByEnum(WsmResourceType.CONTROLLED_AZURE_STORAGE_CONTAINER);
+
+    // save the container definition for downstream steps
+    workingMap.put(
+        WorkspaceFlightMapKeys.ControlledResourceKeys.CLONED_RESOURCE_DEFINITION, clonedContainer);
+
+    var apiCreatedContainer =
+        new ApiCreatedControlledAzureStorageContainer()
+            .azureStorageContainer(clonedContainer.toApiResource())
+            .resourceId(destinationContainerResource.getResourceId());
+
+    var apiContainerResult =
+        new ApiClonedControlledAzureStorageContainer()
```

Comment:
I know it is common practice to build and return the API form of the result here. I want to move us to returning the internal form of the structure and doing the conversion from internal to API form in the controller. We may end up with some additional `record` objects to move the result around, but it will help us better insulate the implementation code and what we serialize in the database from API structures.

This will be part of the clone refactor work, so asking you to change it here is entirely self-serving! :)

Reply:
I've added a [ClonedAzureStorageContainer](https://github.com/DataBiosphere/terra-workspace-manager/blob/arh-wor-532-clone-container-flight-mk2/service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/azure/container/ClonedAzureStorageContainer.java) intermediate model that is set as the flight result here, which the API then adapts to an API result.